infrastructure/src/main/java/org/corfudb/infrastructure/ManagementServer.java

```diff
@@ -203,109 +172,117 @@ private CorfuRuntime getNewCorfuRuntime() {
+                    "incomplete layout {}", req.getHeader().getRequestId(), layout);
+
+            responseHeader = getHeaderMsg(req.getHeader(), false, false);
+            response = getResponseMsg(responseHeader, getBootstrapManagementResponseMsg(false));
+        } else if (serverContext.getManagementLayout() != null) {
+            log.warn("handleBootstrapManagement[{}]: Got a request to bootstrap a server, with layout " +
+                    "{}, which is already bootstrapped... Rejecting!", req.getHeader().getRequestId(), layout);
+
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            response = getResponseMsg(responseHeader, getBootstrappedErrorMsg());
         } else {
-            Layout layout = msg.getPayload();
-            log.info("handleManagementBootstrap: received bootstrap layout : {}", layout);
-            if(layout.getClusterId() == null){
-                log.warn("handleManagementBootstrap: clusterId for the layout {} is not present.",
-                        layout.getClusterId());
-                r.sendResponse(ctx, msg, new CorfuMsg(CorfuMsgType.NACK));
-            }
-            else{
-                serverContext.saveManagementLayout(layout);
-                r.sendResponse(ctx, msg, new CorfuMsg(CorfuMsgType.ACK));
-            }
+            log.info("handleBootstrapManagement[{}]: Received bootstrap " +
+                    "layout {}", req.getHeader().getRequestId(), layout);
+
+            serverContext.saveManagementLayout(layout);
+            responseHeader = getHeaderMsg(req.getHeader(), false, true);
+            response = getResponseMsg(responseHeader, getBootstrapManagementResponseMsg(true));
         }
+
+        r.sendResponse(response, ctx);
     }
 
     /**
      * Triggers the failure handler.
      * The msg contains the failed/defected nodes.
      *
-     * @param msg corfu message containing MANAGEMENT_FAILURE_DETECTED
-     * @param ctx netty ChannelHandlerContext
-     * @param r   server router
+     * @param req  a message containing a REPORT_FAILURE request
+     * @param ctx  the netty ChannelHandlerContext
+     * @param r    the server router
      */
-    @ServerHandler(type = CorfuMsgType.MANAGEMENT_FAILURE_DETECTED)
-    public void handleFailureDetectedMsg(CorfuPayloadMsg<DetectorMsg> msg,
-                                         ChannelHandlerContext ctx, IServerRouter r) {
-
-        // This server has not been bootstrapped yet, ignore all requests.
-        if (!isBootstrapped(msg)) {
-            r.sendResponse(ctx, msg, new CorfuMsg(CorfuMsgType.MANAGEMENT_NOBOOTSTRAP_ERROR));
+    @RequestHandler(type = PayloadCase.REPORT_FAILURE_REQUEST)
+    public void handleReportFailure(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        // If the server isn't bootstrapped yet, ignore the request
+        if (!isBootstrapped(req)) {
+            r.sendNoBootstrapError(req.getHeader(), ctx);
```

Comment:
Do we need to differentiate between a management server no bootstrap error and a non-management server bootstrap error? it seems like we used to have two different errors, but they were essentially the same.

Reply:
I haven't seen a need to. The same exception was thrown in both cases. The only difference now is that this is handled by BaseHandler.